File: apps/mobile/lib/features/announcement/domain/announcement.dart
enum AnnouncementPriority { normal, important, urgent }

/// How an announcement is filed on the noticeboard — a separate question from
/// how urgent it is ([AnnouncementPriority]).
enum AnnouncementCategory { info, event }

/// The chips above the noticeboard.
enum AnnouncementFilter { all, important, event, info }

/// The single badge an announcement carries. Mirrors `announcementBadge()` in
/// `packages/contracts/src/announcement.ts` — urgency wins over filing, so an
/// urgent kegiatan reads as "Penting" first.
enum AnnouncementBadge { important, event, info }

const announcementBadgeLabels = {
  AnnouncementBadge.important: 'Penting',
  AnnouncementBadge.event: 'Acara',
  AnnouncementBadge.info: 'Info',
};

const announcementFilterLabels = {
  AnnouncementFilter.all: 'Semua',
  AnnouncementFilter.important: 'Penting',
  AnnouncementFilter.event: 'Acara',
  AnnouncementFilter.info: 'Info',
};

extension AnnouncementFilterApi on AnnouncementFilter {
  String get apiValue => name;
}

AnnouncementPriority _priorityFromApi(Object? value) {
  switch (value) {
    case 'URGENT':
      return AnnouncementPriority.urgent;
    case 'IMPORTANT':
      return AnnouncementPriority.important;
    default:
      return AnnouncementPriority.normal;
  }
}

AnnouncementCategory _categoryFromApi(Object? value) =>
    value == 'EVENT' ? AnnouncementCategory.event : AnnouncementCategory.info;

AnnouncementBadge _badgeOf(
  AnnouncementPriority priority,
  AnnouncementCategory category,
) {
  if (priority != AnnouncementPriority.normal) return AnnouncementBadge.important;
  return category == AnnouncementCategory.event
      ? AnnouncementBadge.event
      : AnnouncementBadge.info;
}

class AnnouncementSummary {
  const AnnouncementSummary({
    required this.id,
    required this.title,
    required this.summary,
    required this.priority,
    required this.category,
    required this.coverImageUrl,
    required this.publishedAt,
    required this.isRead,
  });

  final String id;
  final String title;
  final String summary;
  final AnnouncementPriority priority;
  final AnnouncementCategory category;
  final String? coverImageUrl;
  final DateTime publishedAt;
  final bool isRead;

  AnnouncementBadge get badge => _badgeOf(priority, category);

  factory AnnouncementSummary.fromJson(Map<String, dynamic> json) {
    return AnnouncementSummary(
      id: json['id'] as String,
      title: json['title'] as String,
      summary: json['summary'] as String,
      priority: _priorityFromApi(json['priority']),
      category: _categoryFromApi(json['category']),
      coverImageUrl: json['coverImageUrl'] as String?,
      publishedAt: DateTime.parse(json['publishedAt'] as String),
      isRead: json['isRead'] as bool? ?? false,
    );
  }
}

class AnnouncementDetail {
  const AnnouncementDetail({
    required this.id,
    required this.title,
    required this.summary,
    required this.body,
    required this.priority,
    required this.category,
    required this.coverImageUrl,
    required this.publishedAt,
    required this.isRead,
  });

  final String id;
  final String title;

  /// The one-or-two sentence blurb shown on the board; the edit form needs it
  /// so saving never silently blanks it.
  final String summary;
  final String body;
  final AnnouncementPriority priority;
  final AnnouncementCategory category;
  final String? coverImageUrl;
  final DateTime publishedAt;
  final bool isRead;

  AnnouncementBadge get badge => _badgeOf(priority, category);

  factory AnnouncementDetail.fromJson(Map<String, dynamic> json) {
    return AnnouncementDetail(
      id: json['id'] as String,
      title: json['title'] as String,
      summary: json['summary'] as String? ?? '',
      body: json['body'] as String,
      priority: _priorityFromApi(json['priority']),
      category: _categoryFromApi(json['category']),
      coverImageUrl: json['coverImageUrl'] as String?,
      publishedAt: DateTime.parse(json['publishedAt'] as String),
      isRead: json['isRead'] as bool? ?? false,
    );
  }
}

File: packages/contracts/src/announcement.ts
import { z } from "zod";

import { dataEnvelopeSchema } from "./envelope";

export const announcementPrioritySchema = z.enum(["NORMAL", "IMPORTANT", "URGENT"]);

/** How an announcement is filed on the noticeboard — a separate question from
 * how urgent it is (`priority`). */
export const announcementCategorySchema = z.enum(["INFO", "EVENT"]);

export type AnnouncementCategory = z.infer<typeof announcementCategorySchema>;

export const announcementSummarySchema = z.object({
  id: z.string().uuid(),
  title: z.string().min(1),
  summary: z.string().min(1),
  priority: announcementPrioritySchema,
  category: announcementCategorySchema,
  coverImageUrl: z.string().url().nullable(),
  publishedAt: z.string().datetime({ offset: true }),
  isRead: z.boolean(),
});

export const announcementDetailSchema = announcementSummarySchema.extend({
  body: z.string().min(1),
});

/** The chips on the noticeboard: everything, the urgent ones, kegiatan, and
 * plain information. */
export const announcementFilterSchema = z.enum(["all", "important", "event", "info"]);

export type AnnouncementFilter = z.infer<typeof announcementFilterSchema>;

export const announcementListQuerySchema = z.object({
  filter: announcementFilterSchema.default("all"),
  limit: z.coerce.number().int().min(1).max(50).default(20),
});

export type AnnouncementListQuery = z.infer<typeof announcementListQuerySchema>;

export type AnnouncementBadge = "important" | "event" | "info";

/**
 * The single badge an announcement carries. Urgency wins over filing: an
 * urgent kegiatan reads as "Penting" first, because that is what a warga
 * scanning the board needs to notice.
 *
 * Deriving it here — rather than storing a third "IMPORTANT" category — is
 * what keeps `priority` and `category` from ever disagreeing about whether
 * something is penting.
 */
export function announcementBadge(announcement: {
  priority: z.infer<typeof announcementPrioritySchema>;
  category: AnnouncementCategory;
}): AnnouncementBadge {
  if (announcement.priority !== "NORMAL") return "important";
  return announcement.category === "EVENT" ? "event" : "info";
}

export const ANNOUNCEMENT_BADGE_LABELS: Record<AnnouncementBadge, string> = {
  important: "Penting",
  event: "Acara",
  info: "Info",
};

export const ANNOUNCEMENT_FILTER_LABELS: Record<AnnouncementFilter, string> = {
  all: "Semua",
  important: "Penting",
  event: "Acara",
  info: "Info",
};

export const announcementListResponseSchema = dataEnvelopeSchema(
  z.object({
    items: z.array(announcementSummarySchema),
  }),
);

export type AnnouncementListResponse = z.infer<typeof announcementListResponseSchema>;

export const announcementDetailResponseSchema = dataEnvelopeSchema(
  z.object({
    announcement: announcementDetailSchema,
  }),
);

export type AnnouncementDetailResponse = z.infer<typeof announcementDetailResponseSchema>;

export const markAnnouncementReadResponseSchema = dataEnvelopeSchema(
  z.object({
    announcementId: z.string().uuid(),
    readAt: z.string().datetime({ offset: true }),
  }),
);

export type MarkAnnouncementReadResponse = z.infer<typeof markAnnouncementReadResponseSchema>;

export type AnnouncementSummary = z.infer<typeof announcementSummarySchema>;
export type AnnouncementDetail = z.infer<typeof announcementDetailSchema>;
export type AnnouncementPriority = z.infer<typeof announcementPrioritySchema>;

export const createAnnouncementSchema = z.object({
  title: z.string().min(3, "Judul minimal 3 karakter").max(240),
  summary: z.string().min(5, "Ringkasan minimal 5 karakter").max(500),
  body: z.string().min(10, "Isi pengumuman minimal 10 karakter"),
  priority: announcementPrioritySchema.default("NORMAL"),
  category: announcementCategorySchema.default("INFO"),
  coverImageUrl: z.string().url().optional(),
});

export type CreateAnnouncementInput = z.infer<typeof createAnnouncementSchema>;

export const updateAnnouncementSchema = z
  .object({
    title: z.string().trim().min(3, "Judul minimal 3 karakter").max(240),
    summary: z.string().trim().min(5, "Ringkasan minimal 5 karakter").max(500),
    body: z.string().trim().min(10, "Isi pengumuman minimal 10 karakter"),
    priority: announcementPrioritySchema,
    category: announcementCategorySchema,
    coverImageUrl: z.string().url().nullable(),
  })
  .partial()
  .refine((input) => Object.keys(input).length > 0, {
    message: "Kirim setidaknya satu perubahan pengumuman.",
  });

export type UpdateAnnouncementInput = z.infer<typeof updateAnnouncementSchema>;

/** Taking a notice off the board archives it rather than erasing the row, so
 * the audit trail and any notification that referenced it stay intact. */
export const archiveAnnouncementResponseSchema = dataEnvelopeSchema(
  z.object({
    announcementId: z.string().uuid(),
    archivedAt: z.string().datetime({ offset: true }),
  }),
);

export type ArchiveAnnouncementResponse = z.infer<typeof archiveAnnouncementResponseSchema>;
